pkg/cmd/adm/install_operator.go

```diff
@@ -0,0 +1,236 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: name.Namespace,
+			Name:      name.Name,
+		},
+		Spec: &olmv1alpha1.SubscriptionSpec{
+			Channel:                "staging",
+			InstallPlanApproval:    olmv1alpha1.ApprovalAutomatic,
+			Package:                operatorName,
+			CatalogSource:          catalogSourceName,
+			CatalogSourceNamespace: name.Namespace,
+		},
+	}
+}
+
+func waitUntilCatalogSourceIsReady(ctx *clicontext.TerminalContext, catalogSourceKey runtimeclient.ObjectKey, waitForReadyTimeout time.Duration) error {
+	cs := &olmv1alpha1.CatalogSource{}
+	if err := wait.PollImmediate(2*time.Second, waitForReadyTimeout, func() (bool, error) {
+		ctx.Printlnf("waiting for CatalogSource %s to become ready", catalogSourceKey)
+		if err := ctx.KubeClient.Get(ctx, catalogSourceKey, cs); err != nil {
+			return false, err
+		}
+
+		return cs.Status.GRPCConnectionState != nil && cs.Status.GRPCConnectionState.LastObservedState == "READY", nil
+	}); err != nil {
+		csString, _ := json.Marshal(cs)
+		return errs.Wrapf(err, "failed waiting for catalog source to be ready.\n CatalogSrouce found: %v \n\t", string(csString))
+	}
+	return nil
+}
+
+func waitUntilInstallPlanIsComplete(ctx *clicontext.TerminalContext, cl runtimeclient.Client, namespace string, waitForReadyTimeout time.Duration) error {
+	plans := &olmv1alpha1.InstallPlanList{}
+	if err := wait.PollImmediate(2*time.Second, waitForReadyTimeout, func() (bool, error) {
+		ctx.Printlnf("waiting for InstallPlans in namespace %s to complete", namespace)
+		if err := cl.List(ctx, plans, runtimeclient.InNamespace(namespace),
+			runtimeclient.MatchingLabels{fmt.Sprintf("operators.coreos.com/%s.%s", namespace, namespace): ""},
+		); err != nil {
+			return false, err
+		}
+
+		for _, ip := range plans.Items {
+			if ip.Status.Phase != olmv1alpha1.InstallPlanPhaseComplete {
+				return false, nil
+			}
+		}
+
+		return len(plans.Items) > 0, nil
+	}); err != nil {
+		plansString, _ := json.Marshal(plans)
+		return errs.Wrapf(err, "failed waiting for install plan to be complete.\n InstallPlans found: %s \n\t", string(plansString))
+	}
+	return nil
+}
+
+// checkOneOperatorPerNamespace returns an error in case the namespace contains the other operator installed.
+// So for host namespace member operator should not be installed in there and vice-versa.
+func checkOneOperatorPerNamespace(ctx *clicontext.TerminalContext, namespace, operator string) error {
+	namespacedName := types.NamespacedName{
+		Namespace: namespace,
+		Name:      getOtherOperator(operator),
+	}
+	subscription := olmv1alpha1.Subscription{}
+	if err := ctx.KubeClient.Get(ctx.Context, namespacedName, &subscription); err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+
+		return err
+	}
+	return fmt.Errorf("found already installed subscription %s in namespace %s", subscription.GetName(), subscription.GetNamespace())
+}
+
+func getOtherOperator(operator string) string {
+	operatorToCheck := map[string]string{
```

Comment:
A `switch` statement would IMHO work here, too, without the need to allocate a new hash map on each invocation.

Reply:
I don't think it's a big performance issue, initially I used the switch but found the hash map nicer. I can put the switch back , no strong opinion on this.